source/client-side-encryption/client-side-encryption.rst

```diff
@@ -752,91 +982,304 @@ See `Why do operations on views fail?`_.
+for encryption or decryption.
 
 Drivers MUST apply timeouts to operations executed as part of client-side encryption per `Client Side Operations
 Timeout: Client Side Encryption
-<../client-side-operations-timeout/client-side-operations-timeout.rst#client-side-encryption>`__.
+<../client-side-operations-timeout/client-side-operations-timeout.rst#client-side-encryption>`_.
 
-Integrating with libmongocrypt
-==============================
+Integrating with `libmongocrypt`
+================================
 
-Each ClientEncryption instance MUST have one handle to libmongocrypt.
-See the `The guide to integrating libmongocrypt <https://github.com/mongodb/libmongocrypt/blob/master/integrating.md>`__
-for more information.
+Each ClientEncryption instance MUST have one `libmongocrypt_handle`.
 
-libmongocrypt exposes logging capabilities. If a driver provides a
+.. seealso::
+
+   `The libmongocrypt C API documentation <lmc-c-api_>`_
+      For information on how to initialize, encrypt, decrypt with libmongocrypt.
+
+   `The Guide to Integrating libmongocrypt <lmc-integrating_>`_
+      For information about integrating the libmongocrypt library in a driver.
+
+`libmongocrypt` exposes logging capabilities. If a driver provides a
 logging mechanism, it MUST enable this logging and integrate. E.g. if
 your driver exposes a logging callback that a user can set, it SHOULD be
 possible to get log messages from libmongocrypt.
 
-Drivers MUST propagate errors from libmongocrypt in whatever way is
+Drivers MUST propagate errors from `libmongocrypt` in whatever way is
 idiomatic to the driver (exception, error object, etc.). These errors
 MUST be distinguished in some way (e.g. exception type) to make it
 easier for users to distinguish when a command fails due to client side
 encryption.
 
+
+.. index:: csfle
+.. _cse.enabling-csfle:
+
+Enabling Command Marking with the `csfle` Library
+=================================================
+
+The MongoDB Enterprise distribution includes a dynamic library named
+``mongo_csfle_v1`` (with the appropriate file extension or filename suffix for
+the host platform). This library will be loaded by `libmongocrypt` when the
+``mongocrypt_init`` function is invoked
+`(from the libmongocrypt C API) <lmc-c-api_>`_ based on the search critera that
+are provided by the driver.
+
+`libmongocrypt` allows the driver to specify an arbitrary list of directory
+paths in which to search for the `csfle` dynamic library. The user may specify
+`csfle` search behavior by specifying options in `extraOptions`: |opt-paths| and
+|opt-path-override|.
+
+.. note::
+
+   The driver MUST NOT manipulate or do any validation on the `csfle` path
+   options provided in `extraOptions`. They should be passed through to
+   `libmongocrypt` unchanged.
+
+
+.. _search path:
+.. _search paths:
+.. _cse.csfle.search-paths:
+
+Setting Search Paths
+--------------------
+
+The driver MUST append `csfle` search paths to the `libmongocrypt_handle`. For
```

Comment:
This open question may be best answered from Java, C#, Python, or Node.JS teams. Those are the four drivers that [package libmongocrypt](https://docs.google.com/spreadsheets/d/1-RcZD3BqxxtiANoM5zeeye5jnkYn-bzqH_VfHXHSxhM/edit#gid=0).

Will drivers that package libmongocrypt need to prepend a custom search path to find `csfle`?

`csfle` has the [enterprise module license](https://github.com/10gen/mongo-enterprise-modules/blob/master/distsrc/LICENSE-Enterprise.txt). Whereas libmongocrypt is Apache licensed. That may motivate a separate package for `csfle`.

Consider [pymongocrypt](https://pypi.org/project/pymongocrypt/). If there is a separate `pymongocrypt-enterprise` package that only wraps `csfle`, then `libmongocrypt` and `csfle` may be installed in separate directories.

Reply:
An option: Another special `"$DEFAULT"` path, which is replaced by the driver's own default, which should be documented. This could also be the sole default value of the search paths array in `extraOptions` (instead of `["$ORIGIN"]`)

Reply:
Is the expectation here that drivers would release a separate package to their package management systems that contains the csfle shared library? If yes, then I think it would be okay to say that a driver MAY add a driver-specific path if it detects the presence of such a package?

Reply:
That's an option, yes. the reason I'm thinking about ``$DEFAULT`` is to allow a user to control whether the driver behaves in this way or not, and to position the `$DEFAULT` in the priority they feels most appropriate for their application. e.g. a user can specify `["some_dir/", "$DEFAULT"]` or `["$DEFAULT", "some_dir/"]` depending on which they prefer.

Reply:
I like the proposed behavior of `$DEFAULT`.